content/blog/foreign-key-constraints.md

```diff
@@ -0,0 +1,198 @@
+
+We can use common table expressions (`WITH` clauses) to do this. We need two queries to get the `conkey` and `confkey` with the corresponding `oid`:
+```
+SELECT oid, unnest(confkey) as confkey
+FROM pg_constraint;
+```
+
+```
+SELECT oid, unnest(conkey) as conkey
+FROM pg_constraint;
+```
+
+Then we `LEFT JOIN` to these expressions in the main query to get a row for every combination of constrained column to referenced column to get the full query:
+
+```
+WITH unnested_confkey AS (
+  SELECT oid, unnest(confkey) as confkey
+  FROM pg_constraint
+),
+unnested_conkey AS (
+  SELECT oid, unnest(conkey) as conkey
+  FROM pg_constraint
+)
+select
+  c.conname                   AS constraint_name,
+  c.contype                   AS constraint_type,
+  tbl.relname                 AS constraint_table,
+  col.attname                 AS constraint_column,
+  referenced_tbl.relname      AS referenced_table,
+  referenced_field.attname    AS referenced_column,
+  pg_get_constraintdef(c.oid) AS definition
+FROM pg_constraint c
+LEFT JOIN unnested_conkey con ON c.oid = con.oid
+LEFT JOIN pg_class tbl ON tbl.oid = c.conrelid
+LEFT JOIN pg_attribute col ON (col.attrelid = tbl.oid AND col.attnum = con.conkey)
+LEFT JOIN pg_class referenced_tbl ON c.confrelid = referenced_tbl.oid
+LEFT JOIN unnested_confkey conf ON c.oid = conf.oid
+LEFT JOIN pg_attribute referenced_field ON (referenced_field.attrelid = c.confrelid AND referenced_field.attnum = conf.confkey)
+WHERE c.contype = 'f';
+```
+
+Which returns the desired result:
+
+```
+   constraint_name   | constraint_type | constraint_table | constraint_column | referenced_table | referenced_column |                  definition                  
+---------------------+-----------------+------------------+-------------------+------------------+-------------------+----------------------------------------------
+ region_parent_fkey  | f               | region           | parent            | region           | name              | FOREIGN KEY (parent) REFERENCES region(name)
+ address_street_fkey | f               | address          | street            | street           | name              | FOREIGN KEY (street) REFERENCES street(name)
+ address_region_fkey | f               | address          | region            | region           | name              | FOREIGN KEY (region) REFERENCES region(name)
+
+
+```
+
+We're restricting the query to only return foreign key constraints but removing the `WHERE` clause will show all constraint types. Care should be taken in interpreting these when constraints are on multiple columns. For example a unique constraint on two separate columns and a multi-column constraint will return similar results, the difference would only be seen in the `constraint_name` and `definition`.  We can demonstrate this by creating two new tables:
+
```

Comment:
More grammar pedantry! Comma before "the difference" should be semi-colon - or split into two sentences.